src/auto_scheduler/compute_dag.cc

```diff
@@ -812,16 +813,24 @@ State ComputeDAG::InferBound(const State& state) const {
 
 Array<State> ComputeDAG::InferBound(const Array<State>& states) const {
   Array<State> out_states;
-  // TODO(jcf94, merrymercy): Use parallel_for to run this in parallel
-  for (const auto& state : states) {
+  out_states.reserve(states.size());
+  std::mutex m;
+
+  support::parallel_for(0, states.size(), [this, &states, &out_states, &m](int i) {
     State out_state;
     try {
-      out_state = this->InferBound(state);
+      out_state = this->InferBound(states[i]);
     } catch (dmlc::Error& e) {
-      LOG(WARNING) << "InferBound fails on the state:\n" << state << "\n" << e.what() << std::endl;
+      LOG(WARNING) << "InferBound fails on the state:\n"
+                   << states[i] << "\n"
+                   << "with: " << e.what() << std::endl;
     }
-    out_states.push_back(std::move(out_state));
-  }
+    if (out_state.defined()) {
+      std::unique_lock<std::mutex> l(m);
```

Comment:
Can we use low-level APIs from ArrayNode to remove this mutex lock?
Because we know the number of states, we can allocate n empty `State()` in advance.

Reply:
Our original implementation in Ansor repo just worked in this way, while if there's any inferbound failure the `out_state` will contains some empy `State()`.

Reply:
Added a note in the doc string of `ComputeDAG::Inferbound`, added necessary `PruneInvalidState` to the search policy.